BUG: Fix bug in AVX-512F np.maximum and np.minimum

r-devulap opened the pull request:
<!-- Please be sure you are following the instructions in the dev guidelines
http://www.numpy.org/devdocs/dev/development_workflow.html
-->

<!-- We'd appreciate it if your commit message is properly formatted
http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message
-->

Fixes https://github.com/numpy/numpy/issues/15597

`np.maximum.accumulate` results in memory overlap for input and output arrays in which case vectorized implementation leads to incorrect results. This patch adds a pre-check condition to avoid running AVX-512F code in case there is a memory overlap.

Commit:
TST: Adding test to validate np.maximum.accumulate and np.minimum.accumulate

Commit:
BUG: Check for memory overlap in AVX-512F implementation of np.maximim and np.minimum

Fixes bug in np.maximum.accumulate and np.minimum.accumulate
See https://github.com/numpy/numpy/issues/15597

r-devulap commented on numpy/core/src/umath/simd.inc.src:
```diff
@@ -85,10 +119,12 @@ abs_ptrdiff(char *a, char *b)
-     (abs(steps[2]) < MAX_STEP_SIZE))
+     (abs(steps[2]) < MAX_STEP_SIZE)  && \
+     (nomemoverlap(args[0], steps[0]*dimensions[0], args[2], steps[2]*dimensions[0])) && \
+     (nomemoverlap(args[1], steps[1]*dimensions[0], args[2], steps[2]*dimensions[0])))
```
This currently ensures that there is no memory overlap across the entire arrays. For large steps and large dimensions, `steps[0]*dimensions[0]` can overflow. If this check an overkill? I don't fully understand under what circumstances does input and output arrays overlap in memory.

r-devulap replied:
For `steps[0]*dimensions[0] `to overflow, `dimensions[0] ` has to be 4398046511104 (since max step size = 2097152 and `npy_intp `has a max value of 2**63 - 1). The memory required to hold such a large array with that step size is 8192 Peta Bytes of data, which I assume is unrealistic.

mattip replied:
code formatting: `steps[0]*dimensions[0]` -> `steps[0] * dimensions[0]`

r-devulap replied:
done.

seberg commented:
I am a bit worried that our tests did not notice this. Could we see if we can add a slow test that checks this type of error for all reduction ufuncs? Or does that even exist, but does not get run on a machine that would notice the problem>

r-devulap commented:
Yeah, I was surprised to see no test coverage for this case. I did add two simple tests to cover `np.maximum.accumulate` and `np.minimum.accumulate`. But would be a good idea to look a bit deeper.

seberg commented on numpy/core/src/umath/simd.inc.src:
```diff
@@ -55,6 +55,40 @@ abs_ptrdiff(char *a, char *b)
 
+/*
+ * nomemoverlap - Do two strided arrays overlap?
+ * Need to accomodate negative strides.
```
Did you accommodate negative strides? The question is a bit tricky, because it used to be the case that they can overlap, however that should not happen anymore (but this is ensured in a completely different part of the code).

Unless my coffee has not kicked in, reductions also have an exact overlap. So basically, it should be sufficient to ensure exact overlapping. But, we should document that somewhere that some inner loops rely on that, and I am not sure where.

seberg replied:
Oh, two strided arrays definitely can overlap, but I think only identically so.

r-devulap replied:
My bad, the function does already accommodate negative strides, I will re-write the comment :)

pv replied:
The loops can be called with overlapping arrays if the data dependence is simple, assuming elementwise operations. This I think currently means in practice that overlap can occur in the reduction/accumulation usage, and when the output pointer never catches the inputs when iterated in order.

Vectorized loops have more strict requirement, so I think they probably should do their own checks.

r-devulap replied:
The function `nomemoverlap `returns a false if the two arrays have overlap. It accommodates for negative strides too. The vectorized loops rely on this function to ensure there is absolutely no overlap between input and output array, so in that sense it is being overly cautious.

seberg commented on numpy/core/src/umath/simd.inc.src:
```diff
@@ -55,6 +55,40 @@ abs_ptrdiff(char *a, char *b)
+{
+    char *ip_start, *ip_end, *op_start, *op_end;
+    if (ip_size < 0)
+    {
```
Just nitpicking, but if you touch the code, we usually put the opening line (except for function defs) on the previousl line.

r-devulap replied:
Ah, damn. I knew this and yet .. I will fix it.

r-devulap commented:
Victim to my own bug :) Hoping my other PR https://github.com/numpy/numpy/pull/15615 fixes the failing CI tests.

Commit:
BUG: Update IS_OUTPUT_BLOCKABLE_UNARY to use the nomemoverlap check

abs_ptrdiff(args[1], args[0]) >= (vsize) does not accomodate strides,
specially when the strides are negative.

mattip commented on numpy/core/src/umath/simd.inc.src:
```diff
@@ -85,19 +116,20 @@ abs_ptrdiff(char *a, char *b)
 /*
  * output should be contiguous, can handle strided input data
  * Input step should be smaller than MAX_STEP_SIZE for performance
  */
```
Comment needs an update for non-overlap condition

r-devulap replied:
Done.

mattip reviewed (commented):
two minor nits, otherwise LGTM.

Commit:
MAINT: Improve formatting and update comments

mattip commented on numpy/core/tests/test_umath.py:
```diff
@@ -3157,6 +3157,13 @@ def test_rint_big_int():
+    out_max = np.array([0.61,0.61,0.77,0.77,0.77], dtype=ftype)
+    out_min = np.array([0.61,0.60,0.60,0.41,0.19], dtype=ftype)
+    assert_equal(np.maximum.accumulate(arr), out_max)
+    assert_equal(np.minimum.accumulate(arr), out_min)
```
Doesn't this only check the non-overlapping case? If so, can you rename it to `test_overlapping_accumulate` and add a case that checks that, I think the original issue provides an instance. Also formatting of the values needs a `' '` after the ','.

r-devulap replied:
`np.maximum.accumulate(arr)` seems to always call `np.maximum` with overlapping arrays. These tests fail the master branch (just like the one provided in the original issue). I fixed the formatting and renamed the test.

mattip commented:
Thanks @r-devulap
